File: web/engine.js
const _INLINE_RE = /\{(\w+)\?([^|{}]*?)(?:\|([^{}]*?))?\}/g;

export function resolveInline(text, state = {}) {
  return text.replace(_INLINE_RE, (_, key, trueBranch, falseBranch) => {
    return state[key] ? trueBranch : (falseBranch ?? "");
  });
}

const _SUBST_RE = /\{(\w+)\}/g;

export function substituteVars(text, state = {}) {
  return text.replace(_SUBST_RE, (match, key) => {
    const val = state[key];
    return val !== undefined && val !== null ? String(val) : match;
  });
}

const _CORRUPT_RE = /\{corrupt(?::([0-9]*\.?[0-9]+))?(?::(consistent|random))?\}([\s\S]*?)\{\/corrupt\}/g;

function _textSeed(text, index) {
  const combined = text + String(index);
  let sum = 0;
  for (let i = 0; i < combined.length; i++) {
    sum = (sum + combined.charCodeAt(i) * (i + 1)) % 4294967296;
  }
  return sum;
}

export function resolveCorruption(text, nodeCorruption) {
  let nodeIntensity = 1.0;
  let nodeMode = "consistent";
  if (typeof nodeCorruption === "number") {
    nodeIntensity = nodeCorruption;
  } else if (nodeCorruption && typeof nodeCorruption === "object") {
    nodeIntensity = nodeCorruption.intensity ?? 1.0;
    nodeMode = nodeCorruption.mode ?? "consistent";
  }

  const segments = [];
  let lastEnd = 0;
  let spanIndex = 0;
  let match;
  _CORRUPT_RE.lastIndex = 0;

  while ((match = _CORRUPT_RE.exec(text)) !== null) {
    if (match.index > lastEnd) {
      segments.push(text.slice(lastEnd, match.index));
    }
    const rawIntensity = match[1];
    const rawMode = match[2];
    const spanText = match[3];
    const intensity = rawIntensity !== undefined ? parseFloat(rawIntensity) : nodeIntensity;
    const mode = rawMode !== undefined ? rawMode : nodeMode;
    const seed = _textSeed(spanText, spanIndex);
    segments.push({ text: spanText, intensity, mode, seed });
    lastEnd = match.index + match[0].length;
    spanIndex++;
  }

  if (lastEnd < text.length) {
    segments.push(text.slice(lastEnd));
  }
  return segments.length ? segments : [text];
}

export function checkRequires(requires = {}, state = {}) {
  return Object.entries(requires).every(([key, condition]) => {
    const current = state[key];
    if (typeof condition === "boolean") {
      return current === condition;
    }
    if (Number.isInteger(condition)) {
      return Number.isInteger(current) && current >= condition;
    }
    if (typeof condition === "string") {
      return current === condition;
    }
    if (Array.isArray(condition)) {
      return condition.includes(current);
    }
    return false;
  });
}

export function applySets(sets = {}, state = {}) {
  const next = { ...state };
  for (const [key, value] of Object.entries(sets)) {
    if (typeof value === "string" && /^[+-]\d+$/.test(value)) {
      const current = Number.isInteger(next[key]) ? next[key] : 0;
      next[key] = current + Number(value);
    } else {
      next[key] = value;
    }
  }
  return next;
}

export function visibleItems(items = [], state = {}) {
  return items.filter((item) => checkRequires(item.requires || {}, state));
}

export function partitionByPosition(items = [], fallback = "after") {
  return {
    before: items.filter((item) => (item.position || fallback) === "before"),
    after: items.filter((item) => (item.position || fallback) === "after"),
  };
}

export function createRun(entry, saved = null, initialState = {}) {
  const story = entry.story;
  return {
    entry,
    story,
    nodeId: saved?.current_node && story.nodes[saved.current_node]
      ? saved.current_node
      : story.meta.start_node,
    history: Array.isArray(saved?.history) ? [...saved.history] : [],
    state: saved?.state && typeof saved.state === "object"
      ? { ...saved.state }
      : { ...initialState },
    currentScene: null,
  };
}

export function currentView(run) {
  const node = run.story.nodes[run.nodeId];
  if (node.scene) {
    run.currentScene = node.scene;
  }
  const choices = visibleItems(node.choices || [], run.state);
  const overlays = partitionByPosition(visibleItems(node.overlays || [], run.state), "after");
  const insets = partitionByPosition(visibleItems(node.insets || [], run.state), "before");
  const pt = (text) => resolveCorruption(
    resolveInline(substituteVars(text, run.state), run.state),
    run.story.nodes[run.nodeId].corruption ?? null
  );
  return {
    node: { ...node, text: pt(node.text) },
    choices,
    overlays: {
      before: overlays.before.map((o) => ({ ...o, text: pt(o.text) })),
      after: overlays.after.map((o) => ({ ...o, text: pt(o.text) })),
    },
    insets: {
      before: insets.before.map((i) => ({ ...i, text: pt(i.text) })),
      after: insets.after.map((i) => ({ ...i, text: pt(i.text) })),
    },
    isEnding: Boolean(node.is_ending || choices.length === 0),
  };
}

export function advance(run, choiceIndex) {
  const view = currentView(run);
  const choice = view.choices[choiceIndex];
  if (!choice) return false;
  run.state = applySets(choice.sets || {}, run.state);
  run.history.push(run.nodeId);
  run.nodeId = choice.next;
  if (run.story.meta.auto_visited_flags !== false) {
    run.state[`visited_${run.nodeId}`] = true;
  }
  return true;
}

File: src/engine.py
from __future__ import annotations

import dataclasses
import re

from .corruption import CorruptedSpan, TextSegments, resolve_corruption
from .display import Display
from .gallery import GalleryManager
from .save import SaveManager, SaveState
from .story import Choice, Story

_DELTA_RE = re.compile(r"^[+-]\d+$")
_INLINE_RE = re.compile(r"\{(\w+)\?([^|{}]*?)(?:\|([^{}]*?))?\}")
_SUBST_RE = re.compile(r"\{(\w+)\}")


class Engine:
    def __init__(
        self,
        story: Story,
        save_manager: SaveManager,
        display: Display,
        gallery_manager: GalleryManager | None = None,
        initial_state: dict[str, bool | int | str] | None = None,
    ) -> None:
        self.story = story
        self.save_manager = save_manager
        self.display = display
        self.gallery_manager = gallery_manager
        self._initial_state: dict[str, bool | int | str] = dict(initial_state) if initial_state else {}
        self._current_node: str = story.start_node
        self._history: list[str] = []
        self._state: dict[str, bool | int | str] = {}
        self._current_scene: str | None = None

    # ------------------------------------------------------------------
    # Public
    # ------------------------------------------------------------------

    def run(self) -> None:
        self._resolve_start()
        self.display.clear_screen()

        while True:
            node = self.story.get_node(self._current_node)
            if node.scene:
                self._current_scene = node.scene

            visible = [c for c in node.choices if self._check_requires(c.requires)]

            visible_overlays = [o for o in node.overlays if self._check_requires(o.requires)]
            before = [o for o in visible_overlays if o.position == "before"]
            after  = [o for o in visible_overlays if o.position == "after"]

            visible_insets = [i for i in node.insets if self._check_requires(i.requires)]
            before_insets = [i for i in visible_insets if i.position == "before"]
            after_insets  = [i for i in visible_insets if i.position == "after"]

            def _pt(text: str) -> TextSegments:
                resolved = self._resolve_inline(
                    self._substitute_vars(text, self._state), self._state
                )
                return resolve_corruption(resolved, node.corruption)

            node_text     = _pt(node.text)
            before_insets = [dataclasses.replace(i, text=_pt(i.text)) for i in before_insets]
            after_insets  = [dataclasses.replace(i, text=_pt(i.text)) for i in after_insets]
            before        = [dataclasses.replace(o, text=_pt(o.text)) for o in before]
            after         = [dataclasses.replace(o, text=_pt(o.text)) for o in after]

            if node.is_ending or not visible:
                self.display.show_ending(node_text, node.ending_type, overlays=before + after)
                if self.gallery_manager:
                    self.gallery_manager.record_ending(self.story.id, self._current_node)
                self.save_manager.delete(self.story.id)
                if self.display.prompt_play_again():
                    self._reset()
                    continue
                return

            self.display.show_node(self.story.title, node_text, before_insets, after_insets, self._current_scene)
            self.display.show_choices(visible, before, after, node.choice_number_color, debug_state=self._state)
            idx = self.display.prompt_choice(visible)
            if idx is None:
                return
            choice = visible[idx - 1]
            self._advance(choice)

    # ------------------------------------------------------------------
    # Private
    # ------------------------------------------------------------------

    def _check_requires(self, requires: dict) -> bool:
        for key, condition in requires.items():
            current = self._state.get(key)
            if isinstance(condition, bool):
                if current != condition:
                    return False
            elif isinstance(condition, int):
                val = current if isinstance(current, int) and not isinstance(current, bool) else 0
                if val < condition:
                    return False
            elif isinstance(condition, str):
                if current != condition:
                    return False
            elif isinstance(condition, list):
                if current not in condition:
                    return False
        return True

    def _apply_sets(self, sets: dict) -> None:
        for key, value in sets.items():
            if isinstance(value, str) and _DELTA_RE.fullmatch(value):
                delta = int(value)
                current = self._state.get(key, 0)
                self._state[key] = (current if isinstance(current, int) and not isinstance(current, bool) else 0) + delta
            else:
                self._state[key] = value

    def _resolve_start(self) -> None:
        if self.save_manager.has_save(self.story.id):
            saved = self.save_manager.load(self.story.id)
            if saved and saved.current_node in self.story.nodes:
                if self.display.prompt_continue_or_new():
                    self._current_node = saved.current_node
                    self._history = saved.history
                    self._state = dict(saved.state)
                    return
        self._current_node = self.story.start_node
        self._history = []
        self._state = dict(self._initial_state)

    def _advance(self, choice: Choice) -> None:
        self._apply_sets(choice.sets)
        self._history.append(self._current_node)
        self._current_node = choice.next
        if self.story.auto_visited_flags:
            self._state[f"visited_{self._current_node}"] = True
        state = SaveState(
            story_id=self.story.id,
            current_node=self._current_node,
            history=list(self._history),
            state=dict(self._state),
        )
        self.save_manager.write(state)
        self.display.show_save_indicator()

    def _reset(self) -> None:
        self._current_node = self.story.start_node
        self._history = []
        self._state = dict(self._initial_state)
        self._current_scene = None
        self.save_manager.delete(self.story.id)

    @staticmethod
    def _substitute_vars(text: str, state: dict) -> str:
        def _replace(m: re.Match) -> str:
            val = state.get(m.group(1))
            return str(val) if val is not None else m.group(0)
        return _SUBST_RE.sub(_replace, text)

    @staticmethod
    def _resolve_inline(text: str, state: dict) -> str:
        def _replace(m: re.Match) -> str:
            true_b = m.group(2)
            false_b = m.group(3) or ""
            return true_b if state.get(m.group(1)) else false_b
        return _INLINE_RE.sub(_replace, text)
